File: android/app/src/main/kotlin/com/example/pdp_todo_app/native/MapPigeonStreamHandler.kt
package com.example.pdp_todo_app.native

import android.os.Handler
import android.os.Looper
import com.example.pdp_todo_app.pigeon.ApiMapClick
import com.example.pdp_todo_app.pigeon.OnMapClickStreamHandler
import com.example.pdp_todo_app.pigeon.PigeonEventSink

class MapPigeonStreamHandler : OnMapClickStreamHandler() {

    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: PigeonEventSink<ApiMapClick>? = null

    override fun onListen(p0: Any?, sink: PigeonEventSink<ApiMapClick>) {
        eventSink = sink
    }

    override fun onCancel(p0: Any?) {
        eventSink = null
    }

    fun emit(latitude: Double, longitude: Double) {
        mainHandler.post {
            eventSink?.success(
                ApiMapClick(
                    latitude = latitude,
                    longitude = longitude,
                )
            )
        }
    }
}

File: ios/Runner/Native/MapPigeonStreamHandler.swift
import Foundation

final class MapPigeonStreamHandler: OnMapClickStreamHandler {
    private var eventSink: PigeonEventSink<ApiMapClick>?

    override func onListen(
        withArguments arguments: Any?,
        sink: PigeonEventSink<ApiMapClick>
    ) {
        eventSink = sink
    }

    override func onCancel(withArguments arguments: Any?) {
        eventSink = nil
    }

    func emit(latitude: Double, longitude: Double) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?.success(
                ApiMapClick(latitude: latitude, longitude: longitude)
            )
        }
    }
}
